src/porepy/numerics/ad/surrogate_operator.py

```diff
@@ -939,64 +906,136 @@ def update_boundary_values(
+        self, domains: Sequence[pp.Grid | pp.MortarGrid], depth: int = 0
+    ) -> None:
+        """Analogous to :meth:`progress_values_in_time`, but for derivatives.
+
+        Important:
+            As of now, there is no use case in storing derivatives in time, except
+            for export and inspection reason. Use with care and do not trash the memory.
+
+        """
+        for grid in domains:
+            data = self._data_of(grid)
+            pp.shift_solution_values(
+                self._name_derivatives, data, pp.TIME_STEP_SOLUTIONS, max_index=depth
+            )
+            current_vals = pp.get_solution_values(
+                self._name_derivatives, data, iterate_index=0
+            )
+            pp.set_solution_values(
+                self._name_derivatives, current_vals, data, time_step_index=0
+            )
+
+    # Methods operating on single grids
+
     def progress_iterate_values_on_grid(
         self,
         values: np.ndarray,
         grid: pp.Grid | pp.MortarGrid,
+        depth: int = 0,
     ) -> None:
         """Shifts the iterate values backwards in the iterate sense and sets ``values``
         as the current iterate values.
 
+        Note:
+            Using this method with ``depth=0`` is a way to set the values of this
+            expression on individual grids, for the current time and iterate.
+
         Parameters:
             values: ``shape=(N,)``
 
                 A new value to be set on the ``grid``. ``N`` is computed from the
                 information passed in ``dof_info`` at instantiation and the number of
                 cells in ``grid`` (and faces and nodes if subdomain).
             grid: A subdomain or interface in the mixed-dimensional domain.
+            depth: ``default=0``
+
+                Maximal number of iterage values stored. The default value leads to
+                **no** previous iterate values stored.
+                See :func:`~porepy.numerics.ad._ad_utils.shift_solution_values`.
 
         Raises:
-            AssertionError: If ``values`` is not of the expected shape.
+            ValueError: If ``values`` is not of the expected shape.
 
         """
         _check_expected_values(
             values, (self.num_dofs_on_grid(grid),), grid, iterate_index=0
         )
         data = self._data_of(grid)
-
-        pp.shift_solution_values(self.name, data, pp.ITERATE_SOLUTIONS)
+        pp.shift_solution_values(self.name, data, pp.ITERATE_SOLUTIONS, max_index=depth)
         pp.set_solution_values(self.name, values, data, iterate_index=0)
 
     def progress_iterate_derivatives_on_grid(
         self,
         values: np.ndarray,
         grid: pp.Grid | pp.MortarGrid,
+        depth: int = 0,
     ) -> None:
         """Analogous to :meth:`progress_iterate_values_on_grid`, but for derivative
         values.
 
+        Important:
+            As of now, there is no use case in storing derivatives in the iterate sense,
```

Comment:
Do we envision such a use case? If not, do we need to keep the method?

Reply:
The method is useful to set the derivative values on **individual** grids, for the current time and iterate (unknown value).
I was thinking about just calling it ``set_derivative_values_on_grid``, for the case we only support derivatives for current time and iterate (like we do with operators as of now).
But the extension to keep derivatives in time and iterate sense is trivial so I included it here for completeness reasons, and chose the default argument such that it essentially only updates the current iterate.

Since you have a far better overview on this topic, feel free to decide whether to keep this or not.

Reply:
Removed functionality: No progress of derivatives in time or iterate sense